### PR TITLE
*: fix bug add partition sometimes do not update placement rules

### LINKED ISSUE #31153: Sometimes add partition will not add placement rules in PD
```
CREATE TABLE tp (id INT) PARTITION BY RANGE (id) followers=1 ( 
        PARTITION p0 VALUES LESS THAN (100),
        PARTITION p1 VALUES LESS THAN (1000)
);

ADD PARTITION (PARTITION p2 VALUES LESS THAN (10000));
```

The new added partition will not have the placement rules synced to PD

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #31153

Problem Summary:

### What is changed and how it works?

When adding a new partition we should recompute the table's basic placement rule including the partition rules that inherited from table. But the current code lost it. It still uses the old table meta. This pr update the table meta with new added partitions and recompute the partition rules

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: ddl/partition.go [hunk 1]
```diff
@@ -187,6 +187,10 @@ func (w *worker) onAddTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (v
 func alterTablePartitionBundles(t *meta.Meta, tblInfo *model.TableInfo, addingDefinitions []model.PartitionDefinition) ([]*placement.Bundle, error) {
 	var bundles []*placement.Bundle
 
+	// tblInfo do not include added partitions, so we should add them first
```
Comment (AilinKid) [inline ddl/partition.go:190]: // append it to tableInfo temporarily to generate the full table-level bundle covering the one with adding-partition range

File: ddl/partition.go [hunk 2]
```diff
@@ -187,6 +187,10 @@ func (w *worker) onAddTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (v
...
+	tblInfo = tblInfo.Clone()
+	tblInfo.Partition.Definitions = append(tblInfo.Partition.Definitions, addingDefinitions...)
```
Comment (AilinKid) [inline ddl/partition.go:192]: seems the test and code are irrelevant

Comment (lcwangchao) [reply ddl/partition.go:192]: It is related in fact. The test will fail if we revert this two line. This bug only occurs when the table has its own placement settings and then we add a new partition without placement (that is to say, inherit from table). So I modified the test case to let the partition table has the placement settings in table level.

**Review by xhebox — APPROVED** [review]
Will a `append` + `defer` to recover the definitions back cheaper? I don't mind much about 'cloning' though.

**Review by AilinKid — APPROVED** [review]

**Review by AilinKid — APPROVED** [review]

**Pull request merged**